libraries/AP_Quicktune/AP_Quicktune.cpp

```diff
@@ -0,0 +1,730 @@
+        default:
+            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+            return Param::END;
+    }
+}
+
+AP_Quicktune::Stage AP_Quicktune::get_stage(AP_Quicktune::Param param)
+{
+    if (param == Param::RLL_P || param == Param::PIT_P || param == Param::YAW_P) {
+        return Stage::P;
+    } else if (param == Param::RLL_I || param == Param::PIT_I || param == Param::YAW_I) {
+        return Stage::I;
+    } else if (param == Param::RLL_D || param == Param::PIT_D || param == Param::YAW_D) {
+        return Stage::D;
+    } else if (param == Param::RLL_SMAX || param == Param::PIT_SMAX || param == Param::YAW_SMAX) {
+        return Stage::SMAX;
+    } else if (param == Param::RLL_FLTT || param == Param::PIT_FLTT || param == Param::YAW_FLTT) {
+        return Stage::FLTT;
+    } else if (param == Param::RLL_FLTD || param == Param::PIT_FLTD || param == Param::YAW_FLTD) {
+        return Stage::FLTD;
+    } else if (param == Param::RLL_FLTE || param == Param::PIT_FLTE || param == Param::YAW_FLTE) {
+        return Stage::FLTE;
+    } else if (param == Param::RLL_FF || param == Param::PIT_FF || param == Param::YAW_FF) {
+        return Stage::FF;
+    } else {
+        INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+        return Stage::END;
+    }
+}
+
+AP_Float *AP_Quicktune::get_param_pointer(AP_Quicktune::Param param)
+{
+    auto &attitude_control = *AC_AttitudeControl::get_singleton();
+    AC_PID* pid_ptr;
+
+    AxisName axis = get_axis(param);
+    switch (axis)
+    {
+        case AxisName::RLL:
+            pid_ptr = &attitude_control.get_rate_roll_pid();
+            break;
+        case AxisName::PIT:
+            pid_ptr =  &attitude_control.get_rate_pitch_pid();
+            break;
+        case AxisName::YAW:
+            pid_ptr =  &attitude_control.get_rate_yaw_pid();
+            break;
+        default:
+            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+            return nullptr;
+    }
+    
+    Stage stage = get_stage(param);
+    switch (stage)
+    {
+        case Stage::P:
+            return &pid_ptr->kP();
+        case Stage::I:
+            return &pid_ptr->kI();
+        case Stage::D:
+            return &pid_ptr->kD();
+        case Stage::SMAX:
+            return &pid_ptr->slew_limit();
+        case Stage::FLTT:
+            return &pid_ptr->filt_T_hz();
+        case Stage::FLTD:
+            return &pid_ptr->filt_D_hz();
+        case Stage::FLTE:
+            return &pid_ptr->filt_E_hz();
+        case Stage::FF:
+            return &pid_ptr->ff();
+        default:
+            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+            return nullptr;
+    }
+}
+
+float AP_Quicktune::get_param_value(AP_Quicktune::Param param)
+{
+    AP_Float *ptr = get_param_pointer(param);
+    if (ptr != nullptr) {
+        return ptr->get();
+    }
+    INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+    return 0.0;
```

Comment:
```suggestion
```

Reply:
![image](https://github.com/user-attachments/assets/977b1ae2-6604-494b-b37a-a06372d7cbf2)
Removing those two lines results in a compilation failure since this is not a void function. And I didn't want to just return zero without noting something's gone wrong, so I put the internal error there. Not sure what else should be done?

Reply:
Sorry, your right, you still need the return. You just don't need the internal error since  `get_param_pointer` will have internal error-ed already.

Reply:
Ah. I've removed this now.